lib/pltr/v2/reducers/root.js

```diff
@@ -39,6 +43,63 @@ const root = (dataRepairers) => (state, action) => {
         nextCardId: nextId(state.cards),
       })
 
+    case ADD_LINES_FROM_TEMPLATE: {
+      // cards from the template need to know the new ids of lines (and sometimes beats) from the template
+      // FOR LINES:
+      // the strategy here is to use the state's next id value + the template id's current value
+      // the card reducer will have access to the state's next id value so it will be able to determine the correct id
+      // FOR BEATS:
+      // cards will use the cardToBeatIdMap to use the current book's beat ids
+      // but if more beats are needed, they will be created with subsequent ids
+      // ALSO:
+      // TODO:
+      // ignore beats unless there are > 1 hierarchyLevels
+      // add cards to leaf nodes (beats) unless there are > 1 hierarchyLevels
+      const bookId = state.ui.currentTimeline
+      let nextIdForBeats = nextBeatId(state.beats)
+      let beatTree = cloneDeep(state.beats[bookId])
+      let createdNewBeats = false
+      // make a card -> beatId mapping (beatId is from existing beats … augmented with new ones)
+      const beatPositionMap = Object.keys(beatTree.index).reduce((acc, id) => {
+        acc[beatTree.index[id].position] = parseInt(id)
+        return acc
+      }, {})
+      const cardToBeatIdMap = action.templateData.cards.reduce((acc, card) => {
+        const beat = action.templateData.beats['1'].index[card.beatId]
```

Comment:
This is `tree.findNode(action.templateData.beats['1'], card.beatId)` as apposed to `.index[card.beatId]`